File: NeetCodeIO/695MaxAreaOfIsland/MaxAreaOfIsland.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        # DFS -> Time:O(4^mn*mn), Space:O(mn)
        ans = 0
        visited = set()

        def dfs(r, c):
            if r >= 0 \
                and c >= 0 \
                and r < len(grid) \
                and c < len(grid[0]) \
                and (r,c) not in visited \
                and grid[r][c] == 1:
                visited.add((r,c))
                return 1 + dfs(r-1, c) + dfs(r, c-1) + dfs(r+1, c) + dfs(r, c+1)
            return 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1 and (i,j) not in visited:
                    area = dfs(i,j)
                    ans = max(ans, area)

        return ans

File: NeetCodeIO/695MaxAreaOfIsland/MaxAreaOfIsland.cpp
class Solution {
public:
    set<pair<int,int>> visited;
    int dfs(int r, int c, vector<vector<int>>& grid) {
        pair<int,int> temp(r, c);
        if (r >= 0 && c >= 0 && r < grid.size() && c < grid[0].size() && visited.find(temp) == visited.end() 
        && grid[r][c] == 1) {
            visited.insert(temp);
            return 1 + dfs(r-1, c, grid) + dfs(r, c-1, grid) + dfs(r, c+1, grid) + dfs(r+1, c, grid);
        }
        return 0;
    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        // DFS -> Time:O(4^mn*mn), Space:O(mn)
        int ans = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                ans = max(ans, dfs(i,j, grid));
            }
        }
        return ans;
    }
};
